// authing-js-sdk/build/umd/lib/testing-helper.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.generateRandomPhone = exports.generateRandomEmail = exports.generateRandomString = exports.getOptionsFromEnv = exports.sleep = exports.randomString = void 0;
    require('dotenv').config({
        path: '.env'
    });
    /** 随机字符串 **/
    exports.randomString = function () {
        return Math.random()
            .toString(36)
            .slice(2);
    };
    /**
     * 睡眠函数
     * @param ms 毫秒
     */
    exports.sleep = function (ms) {
        return new Promise(function (resolve) { return setTimeout(resolve, ms); });
    };
    /**
     * 从环境变量中读取配置
     *
     */
    exports.getOptionsFromEnv = function () {
        return {
            userPoolId: process.env.AUTHING_USERPOOL_ID,
            secret: process.env.AUTHING_USERPOOL_SECRET,
            appId: process.env.AUTHING_APP_ID,
            appHost: process.env.AUTHING_APP_HOST,
            host: process.env.AUTHING_HOST,
            timeout: 10000,
            publicKey: "-----BEGIN PUBLIC KEY-----\nMIGfMA0GCSqGSIb3DQEBAQUAA4GNADCBiQKBgQC4xKeUgQ+Aoz7TLfAfs9+paePb\n5KIofVthEopwrXFkp8OCeocaTHt9ICjTT2QeJh6cZaDaArfZ873GPUn00eOIZ7Ae\n+TiA2BKHbCvloW3w5Lnqm70iSsUi5Fmu9/2+68GZRH9L7Mlh8cFksCicW2Y2W2uM\nGKl64GDcIq3au+aqJQIDAQAB\n-----END PUBLIC KEY-----"
        };
    };
    /**
     * @description 生成随机字符串
     *
     */
    function generateRandomString(length) {
        if (length === void 0) { length = 30; }
        var result = '';
        var characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
        var charactersLength = characters.length;
        for (var i = 0; i < length; i++) {
            result += characters.charAt(Math.floor(Math.random() * charactersLength));
        }
        return result;
    }
    exports.generateRandomString = generateRandomString;
    function generateRandomEmail() {
        return generateRandomString(14) + '@example.com';
    }
    exports.generateRandomEmail = generateRandomEmail;
    /**
     * @description 生成随机手机号
     *
     */
    exports.generateRandomPhone = function () {
        var headerNums = new Array('139', '138', '137', '136', '135', '134', '159', '158', '157', '150', '151', '152', '188', '187', '182', '183', '184', '178', '130', '131', '132', '156', '155', '186', '185', '176', '133', '153', '189', '180', '181', '177');
        var headerNum = headerNums[Math.floor(Math.random() * 10)];
        var bodyNum = Math.random()
            .toString()
            .replace('0.', '')
            .slice(0, 8);
        return headerNum + bodyNum;
    };
});
